Remove remote

dusansimic opened the pull request.

lacymorrow commented:
Yes! Thank you @dusansimic, this brings new life into this project now that future electron versions are supported.

@sindresorhus I understand you don't work much with Electron anymore, would you be willing to look this over or grant another person maintainer access?

dusansimic commented:
Hi @lacymorrow, keep in mind that this pr still has some work left. I'd like to update libraries dependencies (including linter and other test libraries). I just wanted to publish this as a draft pr in case if anyone else is interested in taking a look, especially since this is going to be quite an update for the library so the more eyes we have on this the better.

At the moment I'm trying to get CI to work so it might take some time.

sindresorhus commented:
@dusansimic There has not been any activity on this for a long time. Time to close it? Or do you still intend to finish it some day?

dusansimic commented:
I do plan on finishing it. I just reached a few blockers regarding maintaining the api compatibility I wasn't sure how to solve and it could buried under other things. Most likely course of action I can see is rewriting the API in TS since it's anyway going to break things and people will need to rewrite things in their apps.

It will then be compiled to js + ts definitions before being pushed to npm.

The logic is essentially there, it just needs to be packaged into a sane api. I'll try to do it in the first part of July if that sounds okay?

sindresorhus commented:
> I just reached a few blockers regarding maintaining the api compatibility I wasn't sure how to solve and it could buried under other things

We can make breaking changes where it makes sense.

> Most likely course of action I can see is rewriting the API in TS.

Sure

> I'll try to do it in the first part of July if that sounds okay?

👍

dusansimic commented:
So far the progress is the following:

- [x] Port library to TypeScript and split functions into `main` and `shared` module (main runs only in main process and shared can run in any process).
- [x] Bump some dependencies (ava, electron, typescript, tsd, xo)
- [x] Add some new dependencies (prettier + prettier xo config)
- [x] Port type tests (tsd)
- [x] Port example app
- [ ] Port unit tests (util.platform)
- [ ] Verify npm package will publish correctly

Regarding the unit test (platform function), I wasn't able to port it in ava since I couldn't modify the chrome version property in the process.versions object. I tried to implement the test in playwright but for some reason I had no success with it. I will most likely implement it in the example app so it will require manual testing but it's going to work at least.

sindresorhus commented:
> I wasn't able to port it in ava since I couldn't modify the chrome version property in the process.versions object

Try with `Object.defineProperty`.

dusansimic commented:
> Try with Object.defineProperty.

Thanks, I've tried that and it didn't work for some reason. I tried putting it into the testing context (the `test` function callback) but no luck. I'll still look for solutions for that but for now I've implemented that test in the example app. It should work out of the box when the app is started using `npm run start`.

Other than that, I've changed up various types and worked on cleaning up some code. I'll also update the docs for the functions including the example snippets. I'm not completely sure about some changes so I'll write about them tomorrow.

sindresorhus commented on package.json:
```diff
@@ -11,14 +11,13 @@
-		"start": "electron example.js",
-		"test": "xo && ava && tsd"
+		"start": "electron test/example.cjs",
+		"build": "tsc",
```
Also needs a `"prepare": "npm run build"` so it builds at publish time too.

sindresorhus commented on package.json:
```diff
@@ -36,17 +35,25 @@
+		"@sindresorhus/tsconfig": "^3.0.1",
+		"electron": "^22.0.0",
+		"prettier": "^2.8.8",
+		"prettier-config-xo": "^2.0.0",
```
Just use the `prettier: true` XO setting.

sindresorhus commented on src/main/about-window.ts:
```diff
@@ -0,0 +1,166 @@
+} from 'electron';
+import {is} from '../shared';
+
+type ShowAboutWindowOptions = {
```
It's better to export the types right away than separately at the end.

sindresorhus commented on src/shared/disable-zoom.ts:
```diff
@@ -0,0 +1,26 @@
+import {is} from './is';
+
+/**
+265,276d167
```
⬆️

dusansimic commented:
I added error throws whenever there is a possibility that there is no active window (which is important for functions like [getWindowBoundsCentered](https://github.com/sindresorhus/electron-util/pull/58/files#diff-7b68f09cf8e70ebdc4659e2534f591cfe23c609a2dd179d99a24125cdd3a5eb4R81)). This seemed like the most sane solution to me.

sindresorhus commented on .github/workflows/main.yml:
```diff
@@ -10,12 +10,14 @@ jobs:
         with:
           node-version: ${{ matrix.node-version }}
       - run: npm install
+      - run: npm run build
```
This should be in the `test` run script, not here.

sindresorhus commented:
The readme must be updated.

sindresorhus commented:
We also need to document which API must be called in the main and renderer process.

dusansimic commented:
> We also need to document which API must be called in the main and renderer process.

Do you recommend any docs generators (TS defniitions to docs like those already in readme)?

sindresorhus commented:
> Do you recommend any docs generators (TS defniitions to docs like those already in readme)?

Not really. TSDoc is not very good.

sindresorhus commented:
@dusansimic Friendly bump. It would be great to get this merged. We can leave non-essential things for later.

dusansimic commented:
Sorry for the delay again. Would you consider using a tool like tsdocs.dev for api docs instead of manually maintaining the docs in the repo?
Here is how the page for electron-util currently looks on the website [link](https://tsdocs.dev/docs/electron-util).

lacymorrow commented:
Is there no way to have this work with the renderer route without remote? It would allow supporting Electron <12.

dusansimic commented:
@lacymorrow

The reason for this pr is to introduce support for this module in versions >=12 because at the moment, it's not compatible with them at all. The current latest version of electron-util is supported in Electron versions <12.

I've decided to not use remote module in order to not force it on the users of this module. If users want to access api that is in main process only, they can create ipc channels to get results from those function calls which is what Electron recommends at the moment. It is a relatively big breaking change but it's in line with what Electron recommends.

This has already been summed up quite nicely in the previous pr which didn't go through https://github.com/sindresorhus/electron-util/pull/37#issue-773009157.

It is possible for us to create those ipc channels and emulate behavior of remote but in my opinion, the less *magic* this module does the better.

lacymorrow commented:
To be clear, I'm in favor of deprecating the remote module and limiting the features available in the renderer process in Electron >= 12, however there doesn't seem to be any reason why this new version cannot be backwards-compatible with older versions. Will it still work the same for Electron < 12?

sindresorhus commented:
> Will it still work the same for Electron < 12?

No, the plan is to target a newer Electron version, probably Electron 27.

lacymorrow commented:
I believe by doing something like:
```js
const Electron = require('electron')
if (Electron.remote) {
  // proceed as normal, support Electron < 12
}  
```
or by using `process.versions.electron` it should be possible to have this change remain backwards-compatible.

sindresorhus commented:
It's probably possible to provide backwards compatibility with such an old Electron version, but that's not something I'm interested in providing. It's more than enough work tracking just the lastest-1 of Electron. Electron changes and breaks so much all the time.

sindresorhus commented:
Looking good to me. Thanks for working on this, @dusansimic 🙏
